### PR TITLE
innovation fund post

### PR DESCRIPTION
**Pull request by steveoh**
This needs 

1. a better title. Got any ideas?
   - it seems like an op-ed 
1. A better ending or any ending haha

### CODE REVIEW

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+I think it is a big win to have a skills training and design documented along with the implementation. Alexa knocks this out of the park with its [skill.json](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/skill.template.json) and [models](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/models/en-US.json). Developers can make changes without having to visit a website and deal with UI and UX. The ask CLI from Amazon is awesome and lets you do this and much more. You can deploy your function, update your skill and model, and test your skill. Adding a new question and training it is as simple as editing a json file. Sharing a demo or some code? Everything is documented as text.
+
+[Dialogflow](https://dialogflow.com/) is Google's machine learning natural language processing Alexa equivalent. Currently there is an API for Dialogflow but no supported clients. Everything you want to do has to go through the Dialogflow web console. The models are shared as a zip file that you have to export and import. The content within them is complicated and risky to edit by hand. I did not like having to import a zip file to look at how a demo was implemented.
```
Comment (steveoh) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: It'd be nice to add something about collaboration and source control benefits to skill.json and alexa models. Harder to keep things straight in dialogflow

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+author:
+  display_name: Steve Gourley
+  email: sgourley@utah.gov
+date: 2019-05-28 23:05:51
+categories:
+  - Developer
+tags: []
+---
+
+Developing software for digital assistants is fun. I was lucky enough to be able to spend some time developing the same skill for a Google Assistant and an Amazon Alexa. I am excited to share my opinions on what each platform does well and not so well. To check out the code, visit [agrc/digital-assistant-skills](https://github.com/agrc/digital-assisant-skills) ony GitHub and look t the `alexa-skill` and `google-skill` branches.
```
Comment (stdavis) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
Developing software for digital assistants is fun. I was lucky enough to be able to spend some time developing the same skill for a Google Assistant and an Amazon Alexa. I am excited to share my opinions on what each platform does well and not so well. To check out the code, visit [agrc/digital-assistant-skills](https://github.com/agrc/digital-assisant-skills) on GitHub and look at the `alexa-skill` and `google-skill` branches.
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{: .text-left }
+
+There are so many pages of documentation written for the Google Assistant and Amazon Alexa. It is a massive undertaking. There are multiple versions of the same documentation for each SDK language and then the underlying request response contracts if you decide to roll your own. With that said, I found Alexa's documentation and demos to be better organized and more current. The ecosystem seemed to be treated more as one item. Google seemed to be offering different products that worked together and the documentation did not seem as cohesive. A lot of the demo code was on older versions of their SDK's.
+
+### Real Time Collaboration
+{: .text-left }
+
+Both digital assistants have slack channels.
+
+- `amazonalexa.slack.com`
```
Comment (stdavis) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
- `[amazonalexa.slack.com](https://amazonalexa.slack.com/)`
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+- `googleathome.slack.com`
```
Comment (stdavis) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
- `[googleathome.slack.com](https://googleathome.slack.com/)`
```

**Comment by stdavis** [comment]
This is looking great to me. Do you want to mention the VSCode extension? I don't remember how useful it was.

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+### Location
+{: .text-left }
+
+The Google Assistant and Alexa simulators are very nice. They give you screens to test presentation markup as well as other features. One feature that the Alexa simulator is missing is location. You cannot set a location for the simulator to use. Google makes this very simple and it is on the main UI. A major difference between the two assistants is that addresses are assigned to devices in Amazon. Amazon is thinking, where do I ship you all of your Holiday packages. The Google Assistant is on the every Android phone so location is where am I now.
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
The Google Assistant and Alexa simulators are very nice. They give you screens to test presentation markup as well as other features. One feature that the Alexa simulator is missing is location. You cannot set a location for the simulator to use. Google makes this very simple and the location is on the main UI. A major difference between the two assistants is that addresses are assigned to devices in Amazon. Amazon is thinking, "Where do I ship you all of your holiday packages?" while the Google Assistant is on every Android phone, so location is "Where am I now?"
```

Comment (atribolet) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: Also, it seems like "nice" is used several times in the article? May be worthwhile to consider other terms.

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+## Resources
+{: .text-left }
+
+### Documentation
+{: .text-left }
+
+There are so many pages of documentation written for the Google Assistant and Amazon Alexa. It is a massive undertaking. There are multiple versions of the same documentation for each SDK language and then the underlying request response contracts if you decide to roll your own. With that said, I found Alexa's documentation and demos to be better organized and more current. The ecosystem seemed to be treated more as one item. Google seemed to be offering different products that worked together and the documentation did not seem as cohesive. A lot of the demo code was on older versions of their SDK's.
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
There are so many pages of documentation written for both Google Assistant and Amazon Alexa. It is a massive undertaking. There are multiple versions of the same documentation for each SDK language and then the underlying request response contracts if you decide to roll your own. With that said, I found Alexa's documentation and demos to be better organized and more current. The Alexa ecosystem seemed to be treated more as one item, while Google seemed to be offering different products that worked together, so the documentation did not seem as cohesive. A lot of the Google demo code was on older versions of their SDKs.
```

Comment (atribolet) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: What is a massive undertaking? To read through the documentation? Can we clarify here?

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+### Real Time Collaboration
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
### Real-Time Collaboration
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+{: .text-left }
+
+Both digital assistants have slack channels.
+
+- `[amazonalexa.slack.com](https://amazonalexa.slack.com/)`
+- `[googleathome.slack.com](https://googleathome.slack.com/)`
+
+The Alexa slack channel is awesome. There are a lot of knowledgeable people in there chatting about building skills. Alexa has employees participating in the conversation. The Google channel is the opposite. A lot of questions go unanswered.
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
The Alexa slack channel is awesome. There are a lot of knowledgeable people in there chatting about building skills. Alexa even has employees participating in the conversation. The Google channel is the opposite. A lot of questions go unanswered.
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+Alexa also runs "[office hours](https://www.twitch.tv/amazonalexa)" on twitch. If you have never used twitch, it is becoming a valuable developer relations outreach tool. _It's not only for gaming_. Alexa developers host an hour "show" where they answer questions watchers type in and show off new features. If you stump them, they get back to you later. This is as good as it gets without pair programming with an Amazon employee.
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
Alexa also runs "[office hours](https://www.twitch.tv/amazonalexa)" on twitch. If you have never used twitch, it is becoming a valuable developer relations outreach tool. _It's not only for gaming_. Alexa developers host an hour "show" where they show off new features and answer questions submitted by watchers. If a watcher stumps the developers with a question, the developers get back to the watcher later. This is as good as it gets without pair programming with an Amazon employee.
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+## Wrap up
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
## Wrap-Up
```

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+{: .text-left }
+
+Developing software for digital assistants is fun. Is it the next mobile? I do not think so. Are developers making a lot of money with skills? No. Is one assistant better than the other? I don't think so. I very much enjoyed developing for Alexa but Googles natural language from Dialogflow seemed smarter.
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
Overall, I very much enjoyed developing for Alexa, but Google's natural language from Dialogflow seemed smarter. Is one assistant better than the other? I don't think so. Are digital assistants the next mobile? I do not think so. Are developers making a lot of money with skills? No. But Google Assistant and Amazon Alexa are both great tools, and I enjoyed trying them out.
```

Comment (atribolet) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: So maybe this isn't much better, but I tried! Beginnings and endings of articles are difficult. lol

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+To train Alexa to understand what to do when questions were asked was a bit cumbersome. I created a [website](https://agrc.github.io/digital-assisant-skills/) to generate Alexa utterances. Basically all of the ways a user could phrase a question. Dialogflow seemed a lot smarter and required less training. It certainly did not require a website to generate phrases. Google is winning the natural language processing in my opinion and was much easier to train.
+
+### Beyond Voice
+{: .text-left }
+
+Digital assistants lasted without screens for barely a year. Skills can now play videos, show images, and present content in basic HTML components to compliment the voice responses. Alexa has a presentation language [APL](https://developer.amazon.com/docs/alexa-presentation-language/apl-overview.html) to design for this but had limited SDK support. The Google Actions SDK for nodejs had cards, buttons, tables, images, etc all built in. It was simple and fun to add the visual components while developing. The simulator displays everything nicely.
+
+## Testing
+{: .text-left }
+
```
Comment (atribolet) [inline _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]:
```suggestion
## Testing
```

Comment (atribolet) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: (Sorry, didn't mean to change anything here, but didn't know how to get out of the comment box without deleting my other changes?...)

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
+---
+title: 'A developer comparison of Google Assistant and Amazon Alexa'
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: I was hoping for a more interesting title altogether...

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+author:
+  display_name: Steve Gourley
+  email: sgourley@utah.gov
+date: 2019-05-28 23:05:51
+categories:
+  - Developer
+tags: []
+---
+
+Developing software for digital assistants is fun. I was lucky enough to be able to spend some time developing the same skill for a Google Assistant and an Amazon Alexa. I am excited to share my opinions on what each platform does well and not so well. To check out the code, visit [agrc/digital-assistant-skills](https://github.com/agrc/digital-assisant-skills) on GitHub and look at the `alexa-skill` and `google-skill` branches.
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: yeah i wondered about that as I was writing. I was really the only agrc dev on the project but I like the idea of using we and AGRC

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+## Getting started
+{: .text-left }
+
+We started with the Alexa skill since we had prior personal experience with the ecosystem and were gifted a few devices from [promotions](https://developer.amazon.com/alexa-skills-kit/alexa-developer-skill-promotion) that we could test with.
+
+Getting started with Amazon was a bit of a hurdle since... enterprise government. The hurdle existed since the Utah and Amazon partnership was very young. Authentication with our credential store had not been set up yet. Furthermore, Amazon Web Services and Alexa are two separate entities that do not share credentials. This required a separate developer account for Alexa and an AWS account for lambda, CloudWatch etc. Ironing out all of the IAM permissions to successfully deploy and debug while in an enterprise took some trial and error. With all of that said, the short while that I used my personal account was very smooth.
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: I reworded this whole paragraph. Let's revisit after I get all these changes in.

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+On the other hand, Utah is a Google Drive Enterprise customer. Therefore, getting started with Google Assistant was very simple. No hoops to jump through.
+
+## Development
+{: .text-left }
+
+### Debug Cycle
+{: .text-left }
+
+I chose to use python for the Alexa skill. Partly because the [Alexa skills kit](https://developer.amazon.com/alexa-skills-kit) for [python](https://github.com/alexa/alexa-skills-kit-sdk-for-python) looked really nice when paired with [Flask](http://flask.pocoo.org/) and partly because it is nice to take a break from callbacks and promises once in a while.
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: the javascritp developers will :)

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+It was relatively easy to make Alexa send her requests to my [local flask server](https://github.com/agrc/digital-assisant-skills#development-ceremony) which was basically imitating a lambda function. This was a huge productivity boost as I was able to use the python debugger and all the tools I am already familiar with. [ngrok](https://dashboard.ngrok.com/get-started) makes it all possible by creating a temporary https url to my local machine that I could give to Alexa.
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: i reworded all of this as well. let's revisit

File: _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+I never set up a local server for the Google Assistant. While the Firebase CLI is awesome and you can start a local server, debugging what is happening in that local server is not simple. I could have set up an express server equivalent to flask but I chose not to. In hindsight, I would recommend the express approach since I frustratingly ended up `console.log` debugging in the Firebase logs.
+
+### Skills as Code
+{: .text-left }
+
+I think it is a big win to have a skills training and design documented along with the implementation. Alexa knocks this out of the park with its [skill.json](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/skill.template.json) and [models](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/models/en-US.json). Developers can make changes without having to visit a website and deal with UI and UX. The ask CLI from Amazon is awesome and lets you do this and much more. You can deploy your function, update your skill and model, and test your skill. Adding a new question and training it is as simple as editing a json file. Sharing a demo or some code? Everything is documented as text.
```
[the remark this answers is not in the document]
Comment (steveoh) [reply _posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md]: you train for a question... Yeah this is a bit confusing. I'm trying not to use words like intents and whatnot that are the actual developer jargon

**Pull request merged**